# Movies & Directors Analysis

## Downloading & Importing the files

In [1]:
import pandas as pd
import numpy as np

In [5]:
movies = pd.read_csv("movies.csv", index_col = 0)

In [6]:
directors = pd.read_csv("directors.csv", index_col = 0)

# Basic Data Exploration







## Checking the shape of the dataframes

In [7]:
movies.shape

(1465, 11)

* The Movies dataframe consists of data for **1,465** movies.

In [8]:
directors.shape

(2349, 3)

* The Directors dataframe consists of data for **2,349** directors.

## Extracting first 5 and bottom 5 records from both the dataframes

In [9]:
movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
5                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
5         4767  2007   May   Tuesday

In [10]:
directors.head()

director_name    id gender
0      James Cameron  4762   Male
1     Gore Verbinski  4763   Male
2         Sam Mendes  4764   Male
3  Christopher Nolan  4765   Male
4     Andrew Stanton  4766   Male

In [11]:
movies.tail()

id  budget  popularity  revenue           title  vote_average  \
4736  48363       0           3   321952  The Last Waltz           7.9   
4743  48370   27000          19  3151130          Clerks           7.4   
4748  48375       0           7        0         Rampage           6.0   
4749  48376       0           3        0         Slacker           6.4   
4768  48395  220000          14  2040920     El Mariachi           6.6   

      vote_count  director_id  year month      day  
4736          64         4809  1978   May   Monday  
4743         755         5369  1994   Sep  Tuesday  
4748         131         5148  2009   Aug   Friday  
4749          77         5535  1990   Jul   Friday  
4768         238         5097  1992   Sep   Friday

In [12]:
directors.tail()

director_name    id gender
2344     Shane Carruth  7106   Male
2345  Neill Dela Llana  7107    NaN
2346       Scott Smith  7108    NaN
2347       Daniel Hsia  7109   Male
2348  Brian Herzlinger  7110   Male

## Basic info of the dataframes

In [13]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1465 entries, 0 to 4768
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            1465 non-null   int64  
 1   budget        1465 non-null   int64  
 2   popularity    1465 non-null   int64  
 3   revenue       1465 non-null   int64  
 4   title         1465 non-null   object 
 5   vote_average  1465 non-null   float64
 6   vote_count    1465 non-null   int64  
 7   director_id   1465 non-null   int64  
 8   year          1465 non-null   int64  
 9   month         1465 non-null   object 
 10  day           1465 non-null   object 
dtypes: float64(1), int64(7), object(3)
memory usage: 137.3+ KB


In [14]:
directors.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2349 entries, 0 to 2348
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   director_name  2349 non-null   object
 1   id             2349 non-null   int64 
 2   gender         1724 non-null   object
dtypes: int64(1), object(2)
memory usage: 73.4+ KB


In [15]:
movies.describe()

id        budget   popularity       revenue  vote_average  \
count   1465.000000  1.465000e+03  1465.000000  1.465000e+03   1465.000000   
mean   45225.191126  4.802295e+07    30.855973  1.432539e+08      6.368191   
std     1189.096396  4.935541e+07    34.845214  2.064918e+08      0.818033   
min    43597.000000  0.000000e+00     0.000000  0.000000e+00      3.000000   
25%    44236.000000  1.400000e+07    11.000000  1.738013e+07      5.900000   
50%    45022.000000  3.300000e+07    23.000000  7.578164e+07      6.400000   
75%    45990.000000  6.600000e+07    41.000000  1.792469e+08      6.900000   
max    48395.000000  3.800000e+08   724.000000  2.787965e+09      8.300000   

         vote_count  director_id         year  
count   1465.000000  1465.000000  1465.000000  
mean    1146.396587  5040.192491  2002.615017  
std     1578.077438   258.059631     8.680141  
min        1.000000  4762.000000  1976.000000  
25%      216.000000  4845.000000  1998.000000  
50%      571.000000  4964.000000  2004.000000  
75%     1387.000000  5179.000000  2009.000000  
max    13752.000000  6204.000000  2016.000000

In [17]:
movies.describe(include = "object")

title month     day
count     1465  1465    1465
unique    1465    12       7
top     Avatar   Dec  Friday
freq         1   193     654

In [16]:
directors.describe()

id
count  2349.000000
mean   5936.000000
std     678.242213
min    4762.000000
25%    5349.000000
50%    5936.000000
75%    6523.000000
max    7110.000000

In [18]:
directors.describe(include = "object")

director_name gender
count            2349   1724
unique           2349      2
top     James Cameron   Male
freq                1   1574

## Checking for unique values

In [ ]:
movies.nunique()

id              1465
budget           201
popularity       128
revenue         1267
title           1465
vote_average      51
vote_count      1019
director_id      199
year              41
month             12
day                7
dtype: int64

In [ ]:
directors.nunique()


director_name    2349
id               2349
gender              2
dtype: int64

## Checking for duplicates

In [19]:
duplicates_movies = movies.duplicated()
duplicates_movies

0       False
1       False
2       False
3       False
5       False
        ...  
4736    False
4743    False
4748    False
4749    False
4768    False
Length: 1465, dtype: bool

In [26]:
if duplicates_movies.any():
    print("There are duplicates in the Movies dataframe.")
else:
    print("No duplicates found in the Movies dataframe.")

No duplicates found in the Movies dataframe.


In [22]:
duplicates_directors = directors.duplicated()
duplicates_directors

0       False
1       False
2       False
3       False
4       False
        ...  
2344    False
2345    False
2346    False
2347    False
2348    False
Length: 2349, dtype: bool

In [24]:
if duplicates_directors.any():
    print("There are duplicates in Directors the dataframe.")
else:
    print("No duplicates found in the Directors dataframe.")

No duplicates found in the Directors dataframe.


##  Checking whether all the directors present in the movies dataframe are also present in the directors dataframe.

In [27]:
movies["director_id"].nunique()

199

* There are **199** unique directors are present in the Movies dataframe.

In [28]:
directors["id"].nunique()

2349

* There are **2,349** unique directors are present in the Directors dataframe.

In [29]:
movies["director_id"].isin(directors["id"]).all()

True

* It indicates that all the directors who are present in the Movies dataframe are also present in the Directors dataframe.

## Merging both Dataframes together

In [30]:
df = movies.merge(directors,
                  left_on = "director_id",
                  right_on = "id",
                  how = "left",
                  suffixes = ["_movies","_directors"]
                  )
df

id_movies     budget  popularity     revenue  \
0         43597  237000000         150  2787965087   
1         43598  300000000         139   961000000   
2         43599  245000000         107   880674609   
3         43600  250000000         112  1084939099   
4         43602  258000000         115   890871626   
...         ...        ...         ...         ...   
1460      48363          0           3      321952   
1461      48370      27000          19     3151130   
1462      48375          0           7           0   
1463      48376          0           3           0   
1464      48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name  id_directors  \
0            4762  2009   Dec  Thursday      James Cameron          4762   
1            4763  2007   May  Saturday     Gore Verbinski          4763   
2            4764  2015   Oct    Monday         Sam Mendes          4764   
3            4765  2012   Jul    Monday  Christopher Nolan          4765   
4            4767  2007   May   Tuesday          Sam Raimi          4767   
...           ...   ...   ...       ...                ...           ...   
1460         4809  1978   May    Monday    Martin Scorsese          4809   
1461         5369  1994   Sep   Tuesday        Kevin Smith          5369   
1462         5148  2009   Aug    Friday           Uwe Boll          5148   
1463         5535  1990   Jul    Friday  Richard Linklater          5535   
1464         5097  1992   Sep    Friday   Robert Rodriguez          5097   

     gender  
0      Male  
1      Male  
2      Male  
3      Male  
4      Male  
...     ...  
1460   Male  
1461   Male  
1462   Male  
1463   Male  
1464    NaN  

[1465 rows x 14 columns]

## Dropping unnecessary columns

In [31]:
df.drop(columns = ["director_id","id_directors"], inplace = True)

# Since we have director's names, we don't need their IDs

In [32]:
df.head()

id_movies     budget  popularity     revenue  \
0      43597  237000000         150  2787965087   
1      43598  300000000         139   961000000   
2      43599  245000000         107   880674609   
3      43600  250000000         112  1084939099   
4      43602  258000000         115   890871626   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name gender  
0   Dec  Thursday      James Cameron   Male  
1   May  Saturday     Gore Verbinski   Male  
2   Oct    Monday         Sam Mendes   Male  
3   Jul    Monday  Christopher Nolan   Male  
4   May   Tuesday          Sam Raimi   Male

## Renaming the column

In [33]:
df.rename(columns = {"id_movies":"movies_id"}, inplace = True)

In [34]:
df.head()

movies_id     budget  popularity     revenue  \
0      43597  237000000         150  2787965087   
1      43598  300000000         139   961000000   
2      43599  245000000         107   880674609   
3      43600  250000000         112  1084939099   
4      43602  258000000         115   890871626   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name gender  
0   Dec  Thursday      James Cameron   Male  
1   May  Saturday     Gore Verbinski   Male  
2   Oct    Monday         Sam Mendes   Male  
3   Jul    Monday  Christopher Nolan   Male  
4   May   Tuesday          Sam Raimi   Male

## Converting budget and revenue into millions

In [35]:
df["budget"] = round(df["budget"]/1000000,2)

In [36]:
df["revenue"] = round(df["revenue"]/1000000,2)

In [37]:
df.head()

movies_id  budget  popularity  revenue  \
0      43597   237.0         150  2787.97   
1      43598   300.0         139   961.00   
2      43599   245.0         107   880.67   
3      43600   250.0         112  1084.94   
4      43602   258.0         115   890.87   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name gender  
0   Dec  Thursday      James Cameron   Male  
1   May  Saturday     Gore Verbinski   Male  
2   Oct    Monday         Sam Mendes   Male  
3   Jul    Monday  Christopher Nolan   Male  
4   May   Tuesday          Sam Raimi   Male

In [38]:
df.rename(columns = {"budget":"budget_mil","revenue":"revenue_mil"}, inplace = True)

In [39]:
df.head() # Now we have a nice and clean data for analysis

movies_id  budget_mil  popularity  revenue_mil  \
0      43597       237.0         150      2787.97   
1      43598       300.0         139       961.00   
2      43599       245.0         107       880.67   
3      43600       250.0         112      1084.94   
4      43602       258.0         115       890.87   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name gender  
0   Dec  Thursday      James Cameron   Male  
1   May  Saturday     Gore Verbinski   Male  
2   Oct    Monday         Sam Mendes   Male  
3   Jul    Monday  Christopher Nolan   Male  
4   May   Tuesday          Sam Raimi   Male

## Identify and list the top 5 movies with the highest popularity scores.

In [ ]:
df.sort_values(by = "popularity", ascending = False)[["title","popularity"]].head()

title  popularity
58                                        Interstellar         724
78                                  Mad Max: Fury Road         434
119  Pirates of the Caribbean: The Curse of the Bla...         271
120              The Hunger Games: Mockingjay - Part 1         206
45                                     The Dark Knight         187

* Among all the top 5 movies **Interstellar** stands out to be a highly populated movie.

## Find and list the movies with the highest vote average ratings.

highly rated movies = vote_average > 7

In [44]:
df.loc[df["vote_average"] > 7,["title","vote_average"]].sort_values(
                                    by = "vote_average" ,ascending = False)

title  vote_average
1284                      Pulp Fiction           8.3
901                   Schindler's List           8.3
383                         Fight Club           8.3
917                         GoodFellas           8.2
45                     The Dark Knight           8.2
...                                ...           ...
1344                              Race           7.1
506                    Erin Brockovich           7.1
388   Perfume: The Story of a Murderer           7.1
1371                         They Live           7.1
625                         The Intern           7.1

[301 rows x 2 columns]

* Among 1,465 movies, **301** movies stand out as the highly rated movies.

* Among all the highly rated movies **Pulp Fiction** followed by **Schindler's List** & **Fight Club** stands out as the most highly rated movies.

## Calculate the number of movies released in each year.

In [62]:
df["year"].min()

1976

In [63]:
df["year"].max()

2016

In [55]:
df.groupby(["year"]).agg(
    no_of_movies = ("title","count")).sort_values(
        by = "year", ascending = False).reset_index().head(10)

# Number of movies released each year

year  no_of_movies
0  2016            26
1  2015            41
2  2014            53
3  2013            57
4  2012            49
5  2011            67
6  2010            63
7  2009            66
8  2008            67
9  2007            54

In [59]:
df.groupby(["year"]).agg(
    no_of_movies = ("title","count")).sort_values(
        by = "no_of_movies", ascending = False).reset_index().head()

year  no_of_movies
0  2006            73
1  2005            71
2  2002            69
3  2011            67
4  2008            67

* From the year 1976 to 2016, the years **2006**, **2005**, and **2002** had the **maximum** number of movies released.

In [60]:
df.groupby(["year"]).agg(
    no_of_movies = ("title","count")).sort_values(
        by = "no_of_movies", ascending = True).reset_index().head()

year  no_of_movies
0  1976             3
1  1979             4
2  1978             5
3  1977             6
4  1980             8

* From the year 1976 to 2016, the years **1976**, **1979**, and **1978** had the **minimum** number of movies released.

## Identify directors who have the highest number of movies in the dataset.

In [64]:
df.groupby(["director_name"]).agg(
    no_of_movies = ("title","count")).sort_values(
       by = "no_of_movies", ascending = False).reset_index().head()

director_name  no_of_movies
0  Steven Spielberg            26
1    Clint Eastwood            19
2   Martin Scorsese            19
3       Woody Allen            18
4  Robert Rodriguez            16

* Among 199 directors, **Steven Spielberg**, **Clint Eastwood**, and **Martin Scorsese** are the directors who released the most number of movies, with **Spielberg** leading the list.

## Find directors known for producing high-budget movies by analyzing the budget of movies they've directed.

In [ ]:
df.sort_values(by = "budget_mil", ascending = False)[["director_name","budget_mil"]].head()

director_name  budget_mil
12    Rob Marshall       380.0
1   Gore Verbinski       300.0
6     Bryan Singer       270.0
4        Sam Raimi       258.0
9   Gore Verbinski       255.0

* Among 199 directors, **Rob Marshall**, **Gore Verbinski**, and **Bryan Singer** are the directors known for producing high-budget movies.

## find out latest highly rated movies

latest--> after 2014

highly rated movies --> 7+

In [ ]:
df.loc[(df["year"]>2014) & (df["vote_average"]>7)]

movies_id  budget_mil  popularity  revenue_mil                    title  \
30        43641       190.0         102      1506.25                Furious 7   
78        43724       150.0         434       378.86       Mad Max: Fury Road   
106       43773       135.0         100       532.95             The Revenant   
162       43867       108.0         167       630.16              The Martian   
312       44128        75.0          48       108.15  The Man from U.N.C.L.E.   
394       44281        44.0          68       155.76        The Hateful Eight   
625       44770        35.0          53       194.56               The Intern   
635       44784        40.0          48       165.48          Bridge of Spies   
808       45194        30.0          65        91.71                 Southpaw   
833       45293        28.0          61       201.63   Straight Outta Compton   
839       45301        28.0          57       133.35            The Big Short   
1344      47181         5.0          22        24.80                     Race   

      vote_average  vote_count  year month        day  \
30             7.3        4176  2015   Apr  Wednesday   
78             7.2        9427  2015   May  Wednesday   
106            7.3        6396  2015   Dec     Friday   
162            7.6        7268  2015   Sep  Wednesday   
312            7.1        2265  2015   Aug   Thursday   
394            7.6        4274  2015   Dec     Friday   
625            7.1        1881  2015   Sep   Thursday   
635            7.2        2583  2015   Oct   Thursday   
808            7.3        2067  2015   Jun     Monday   
833            7.7        1355  2015   Aug   Thursday   
839            7.3        2607  2015   Dec     Friday   
1344           7.1         478  2016   Feb     Friday   

                    director_name  gender  
30                      James Wan    Male  
78                  George Miller    Male  
106   Alejandro González Iñárritu    Male  
162                  Ridley Scott    Male  
312                   Guy Ritchie    Male  
394             Quentin Tarantino    Male  
625                  Nancy Meyers  Female  
635              Steven Spielberg    Male  
808                 Antoine Fuqua    Male  
833                  F. Gary Gray    Male  
839                    Adam McKay    Male  
1344              Stephen Hopkins    Male

## Calculate profit for each movie

In [65]:
def calculate_profit(x):
  return x["revenue_mil"]-x["budget_mil"]

In [66]:
df["profit"] = df[["revenue_mil","budget_mil"]].apply(calculate_profit, axis = 1)

In [67]:
df.head()

movies_id  budget_mil  popularity  revenue_mil  \
0      43597       237.0         150      2787.97   
1      43598       300.0         139       961.00   
2      43599       245.0         107       880.67   
3      43600       250.0         112      1084.94   
4      43602       258.0         115       890.87   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name gender   profit  
0   Dec  Thursday      James Cameron   Male  2550.97  
1   May  Saturday     Gore Verbinski   Male   661.00  
2   Oct    Monday         Sam Mendes   Male   635.67  
3   Jul    Monday  Christopher Nolan   Male   834.94  
4   May   Tuesday          Sam Raimi   Male   632.87

## Most 5 profitable movies

In [ ]:
df.sort_values(by = "profit", ascending = False)[["title","profit"]].head() # Here Profit is also in million

title   profit
0                                           Avatar  2550.97
19                                         Titanic  1645.03
30                                       Furious 7  1316.25
199  The Lord of the Rings: The Return of the King  1024.89
36                  Transformers: Dark of the Moon   928.75

* Among the top 5 profitable movies, **Avatar** stands out as the most profitable movie, with a profit of **2,551 million**.

## find out all the highest rated movies year wise

In [69]:
highly_rated_movies = df.sort_values(by = ["year","vote_average"], ascending = [True,False])

In [71]:
highly_rated_movies.drop_duplicates(subset = ["year"], keep = "first", inplace = True)

In [74]:
highly_rated_movies[["year","title","vote_average",]] # highly rated movies year wise

year                                              title  vote_average
1427  1976                                        Taxi Driver           8.0
1368  1977                                         Annie Hall           7.8
1460  1978                                     The Last Waltz           7.9
772   1979                                     Apocalypse Now           8.0
1340  1980                                   The Elephant Man           7.9
1216  1981                                           Das Boot           7.9
856   1982                                       Blade Runner           7.9
919   1983                                           Scarface           8.0
1320  1984                                     The Terminator           7.3
1061  1985                                 Back to the Future           8.0
1286  1986                                        Stand by Me           7.8
905   1987                                   The Untouchables           7.6
898   1988                                           Rain Man           7.6
552   1989                 Indiana Jones and the Last Crusade           7.6
917   1990                                         GoodFellas           8.2
1000  1991                           The Silence of the Lambs           8.1
1426  1992                                     Reservoir Dogs           8.0
901   1993                                   Schindler's List           8.3
1284  1994                                       Pulp Fiction           8.3
785   1995                                              Se7en           8.1
1385  1996                                      Trainspotting           7.8
1240  1997                                  Good Will Hunting           7.9
360   1998                                Saving Private Ryan           7.9
383   1999                                         Fight Club           8.3
1341  2000                                            Memento           8.1
156   2001  The Lord of the Rings: The Fellowship of the Ring           8.0
200   2002              The Lord of the Rings: The Two Towers           8.0
199   2003      The Lord of the Rings: The Return of the King           8.1
787   2004                                Million Dollar Baby           7.7
855   2005                                  Pride & Prejudice           7.7
641   2006                                       The Prestige           8.0
921   2007                                There Will Be Blood           7.9
45    2008                                    The Dark Knight           8.2
332   2009                               Inglourious Basterds           7.9
59    2010                                          Inception           8.1
939   2011                                            Warrior           7.7
178   2012                                   Django Unchained           7.8
183   2013                            The Wolf of Wall Street           7.9
58    2014                                       Interstellar           8.1
833   2015                             Straight Outta Compton           7.7
1344  2016                                               Race           7.1

## Get all the records where director is James Cameron

In [ ]:
df.loc[df["director_name"] == "James Cameron"]

movies_id  budget_mil  popularity  revenue_mil  \
0         43597       237.0         150      2787.97   
19        43622       200.0         100      1845.03   
170       43876       100.0         101       520.00   
173       43879       115.0          38       378.88   
344       44184        70.0          24        90.00   
1100      46000        18.5          67       183.32   
1320      47036         6.4          74        78.37   

                           title  vote_average  vote_count  year month  \
0                         Avatar           7.2       11800  2009   Dec   
19                       Titanic           7.5        7562  1997   Nov   
170   Terminator 2: Judgment Day           7.7        4185  1991   Jul   
173                    True Lies           6.8        1116  1994   Jul   
344                    The Abyss           7.1         808  1989   Aug   
1100                      Aliens           7.7        3220  1986   Jul   
1320              The Terminator           7.3        4128  1984   Oct   

            day  director_name gender   profit   profit  
0      Thursday  James Cameron   Male  2550.97  2550.97  
19      Tuesday  James Cameron   Male  1645.03  1645.03  
170      Monday  James Cameron   Male   420.00   420.00  
173    Thursday  James Cameron   Male   263.88   263.88  
344   Wednesday  James Cameron   Male    20.00    20.00  
1100     Friday  James Cameron   Male   164.82   164.82  
1320     Friday  James Cameron   Male    71.97    71.97

## Calculate the career span and productivity rate for the same director


Career span: The total number of years they have devoted to the industry.

Productivity Rate: Calculated as the total number of movies released by the director divided by career span.

In [ ]:
df_jc = df.loc[df["director_name"] == "James Cameron"]
df_jc

movies_id  budget_mil  popularity  revenue_mil  \
0         43597       237.0         150      2787.97   
19        43622       200.0         100      1845.03   
170       43876       100.0         101       520.00   
173       43879       115.0          38       378.88   
344       44184        70.0          24        90.00   
1100      46000        18.5          67       183.32   
1320      47036         6.4          74        78.37   

                           title  vote_average  vote_count  year month  \
0                         Avatar           7.2       11800  2009   Dec   
19                       Titanic           7.5        7562  1997   Nov   
170   Terminator 2: Judgment Day           7.7        4185  1991   Jul   
173                    True Lies           6.8        1116  1994   Jul   
344                    The Abyss           7.1         808  1989   Aug   
1100                      Aliens           7.7        3220  1986   Jul   
1320              The Terminator           7.3        4128  1984   Oct   

            day  director_name gender   profit   profit  
0      Thursday  James Cameron   Male  2550.97  2550.97  
19      Tuesday  James Cameron   Male  1645.03  1645.03  
170      Monday  James Cameron   Male   420.00   420.00  
173    Thursday  James Cameron   Male   263.88   263.88  
344   Wednesday  James Cameron   Male    20.00    20.00  
1100     Friday  James Cameron   Male   164.82   164.82  
1320     Friday  James Cameron   Male    71.97    71.97

In [ ]:
df_jc["year"].max()

2009

In [ ]:
df_jc["year"].min()

1984

In [ ]:
career_span = df_jc["year"].max() - df_jc["year"].min()
career_span

25

In [ ]:
df_jc["title"].nunique()

7

In [ ]:
productivity = df_jc["title"].nunique()/career_span
productivity

0.28

* **James Cameron** has a **25-year** career span with **0.28** productivity rate.

## Calculate career span and productivity rate for all the directors

In [75]:
df.groupby(["director_name"]).ngroups # Total no. of unique directors present in the dataframe

199

In [76]:
df.groupby(["director_name"]).groups # All the directors with their set of index

{'Adam McKay': [176, 323, 366, 505, 839, 916], 'Adam Shankman': [265, 300, 350, 404, 458, 843, 999, 1231], 'Alejandro González Iñárritu': [106, 749, 1015, 1034, 1077, 1405], 'Alex Proyas': [95, 159, 514, 671, 873], 'Alexander Payne': [793, 1006, 1101, 1211, 1281], 'Andrew Adamson': [11, 43, 328, 501, 947], 'Andrew Niccol': [533, 603, 701, 722, 1439], 'Andrzej Bartkowiak': [349, 549, 754, 911, 924], 'Andy Fickman': [517, 681, 909, 926, 973, 1023], 'Andy Tennant': [314, 320, 464, 593, 676, 885], 'Ang Lee': [99, 134, 748, 840, 1089, 1110, 1132, 1184], 'Anne Fletcher': [610, 650, 736, 789, 1206], 'Antoine Fuqua': [310, 338, 424, 467, 576, 808, 818, 1105], 'Atom Egoyan': [946, 1128, 1164, 1194, 1347, 1416], 'Barry Levinson': [313, 319, 471, 594, 878, 898, 1013, 1037, 1082, 1143, 1185, 1345, 1378], 'Barry Sonnenfeld': [13, 48, 90, 205, 591, 778, 783], 'Ben Stiller': [209, 212, 547, 562, 850], 'Bill Condon': [102, 307, 902, 1233, 1381], 'Bobby Farrelly': [352, 356, 481, 498, 624, 630, 654, 80

In [82]:
directors_productivity = df.groupby(["director_name"]).agg(year_max = ("year","max"),
                                    year_min = ("year", "min"),title_count = ("title","count")).reset_index()
directors_productivity

director_name  year_max  year_min  title_count
0                     Adam McKay      2015      2004            6
1                  Adam Shankman      2012      2001            8
2    Alejandro González Iñárritu      2015      2000            6
3                    Alex Proyas      2016      1994            5
4                Alexander Payne      2013      1999            5
..                           ...       ...       ...          ...
194                   Wes Craven      2011      1984           10
195            Wolfgang Petersen      2006      1981            7
196                  Woody Allen      2013      1977           18
197                  Zack Snyder      2016      2004            7
198                  Zhang Yimou      2014      2002            6

[199 rows x 4 columns]

In [85]:
directors_productivity["career_span"] = directors_productivity["year_max"] - directors_productivity["year_min"]
directors_productivity["productivity_rate"] = directors_productivity["title_count"]/directors_productivity["career_span"]

In [86]:
directors_productivity

director_name  year_max  year_min  title_count  \
0                     Adam McKay      2015      2004            6   
1                  Adam Shankman      2012      2001            8   
2    Alejandro González Iñárritu      2015      2000            6   
3                    Alex Proyas      2016      1994            5   
4                Alexander Payne      2013      1999            5   
..                           ...       ...       ...          ...   
194                   Wes Craven      2011      1984           10   
195            Wolfgang Petersen      2006      1981            7   
196                  Woody Allen      2013      1977           18   
197                  Zack Snyder      2016      2004            7   
198                  Zhang Yimou      2014      2002            6   

     career_span  productivity_rate  
0             11           0.545455  
1             11           0.727273  
2             15           0.400000  
3             22           0.227273  
4             14           0.357143  
..           ...                ...  
194           27           0.370370  
195           25           0.280000  
196           36           0.500000  
197           12           0.583333  
198           12           0.500000  

[199 rows x 6 columns]

In [87]:
directors_productivity.sort_values(by = "productivity_rate", ascending = False)

# High productivity_rate directors

director_name  year_max  year_min  title_count  career_span  \
190       Tyler Perry      2013      2006            9            7   
73    Jason Friedberg      2010      2006            5            4   
169        Shawn Levy      2014      2002           11           12   
158  Robert Rodriguez      2014      1992           16           22   
1       Adam Shankman      2012      2001            8           11   
..                ...       ...       ...          ...          ...   
104   Lawrence Kasdan      2012      1985            5           27   
109        Luc Besson      2014      1985            5           29   
157    Robert Redford      2010      1980            5           30   
170      Sidney Lumet      2006      1976            5           30   
117     Michael Apted      2010      1980            5           30   

     productivity_rate  
190           1.285714  
73            1.250000  
169           0.916667  
158           0.727273  
1             0.727273  
..                 ...  
104           0.185185  
109           0.172414  
157           0.166667  
170           0.166667  
117           0.166667  

[199 rows x 6 columns]

## find out all the movies created by high budgedirector

 high budget director --> he has created a movie of budget >=100 Mil $ atleast once in his life

In [ ]:
high_budget_directors = df.groupby(["director_name"]).filter(lambda x: x["budget_mil"].max()>= 100)

In [ ]:
high_budget_directors

movies_id  budget_mil  popularity  revenue_mil  \
0         43597      237.00         150      2787.97   
1         43598      300.00         139       961.00   
2         43599      245.00         107       880.67   
3         43600      250.00         112      1084.94   
4         43602      258.00         115       890.87   
...         ...         ...         ...          ...   
1450      48267        0.40          33       100.00   
1451      48268        0.20          13         4.51   
1452      48274        0.00           5         2.61   
1458      48335        0.06          27         3.22   
1460      48363        0.00           3         0.32   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1450                                   Mad Max           6.6        1213   
1451                                  Swingers           6.8         253   
1452                                     Three           6.3          31   
1458                                        Pi           7.1         586   
1460                            The Last Waltz           7.9          64   

      year month       day      director_name gender   profit  
0     2009   Dec  Thursday      James Cameron   Male  2550.97  
1     2007   May  Saturday     Gore Verbinski   Male   661.00  
2     2015   Oct    Monday         Sam Mendes   Male   635.67  
3     2012   Jul    Monday  Christopher Nolan   Male   834.94  
4     2007   May   Tuesday          Sam Raimi   Male   632.87  
...    ...   ...       ...                ...    ...      ...  
1450  1979   Apr  Thursday      George Miller   Male    99.60  
1451  1996   Oct    Friday         Doug Liman   Male     4.31  
1452  2010   Dec  Thursday         Tom Tykwer   Male     2.61  
1458  1998   Jul    Friday   Darren Aronofsky   Male     3.16  
1460  1978   May    Monday    Martin Scorsese   Male     0.32  

[679 rows x 13 columns]

In [ ]:
high_budget_directors["director_name"].unique()

# These are the all unique high-budget directors

array(['James Cameron', 'Gore Verbinski', 'Sam Mendes',
       'Christopher Nolan', 'Sam Raimi', 'Zack Snyder', 'Bryan Singer',
       'Marc Forster', 'Andrew Adamson', 'Rob Marshall',
       'Barry Sonnenfeld', 'Peter Jackson', 'Ridley Scott', 'Chris Weitz',
       'Peter Berg', 'Tim Burton', 'Brett Ratner', 'Michael Bay',
       'Martin Campbell', 'McG', 'James Wan', 'Mike Newell',
       'Guillermo del Toro', 'Steven Spielberg', 'Justin Lin',
       'Roland Emmerich', 'Robert Zemeckis', 'Lilly Wachowski',
       'Jon Favreau', 'Martin Scorsese', 'Rob Cohen', 'David Ayer',
       'Tom Shadyac', 'Doug Liman', 'Kevin Reynolds', 'David Fincher',
       'Francis Lawrence', 'Jon Turteltaub', 'Wolfgang Petersen',
       'Michael Apted', 'Oliver Stone', 'Shawn Levy', 'George Miller',
       'Ron Howard', 'Kenneth Branagh', 'Jonathan Liebesman',
       'M. Night Shyamalan', 'Joe Wright', 'Rob Minkoff', 'Lee Tamahori',
       'Edward Zwick', 'Alex Proyas', 'Richard Donner', 'Ang Lee',
       

In [ ]:
high_budget_directors["director_name"].nunique() # Total number of unique high-budget directors

85

In [ ]:
df["director_name"].nunique() # Total number of unique directors

199

* Out of 199 directors, **85** directors stand out as the high-budget directors.

## Calculate all the risky movies

risky movies = budget > revenue

In [ ]:
def calculate_risky_movies(x):
  x["risky"] = (x["budget_mil"]-x["revenue_mil"].mean())>=0
  return x

In [ ]:
risky_movies = df.groupby(["director_name"]).apply(calculate_risky_movies)

In [ ]:
risky_movies

movies_id  budget_mil  popularity  revenue_mil  \
0         43597      237.00         150      2787.97   
1         43598      300.00         139       961.00   
2         43599      245.00         107       880.67   
3         43600      250.00         112      1084.94   
4         43602      258.00         115       890.87   
...         ...         ...         ...          ...   
1460      48363        0.00           3         0.32   
1461      48370        0.03          19         3.15   
1462      48375        0.00           7         0.00   
1463      48376        0.00           3         0.00   
1464      48395        0.22          14         2.04   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender   profit  risky  
0     2009   Dec  Thursday      James Cameron   Male  2550.97  False  
1     2007   May  Saturday     Gore Verbinski   Male   661.00  False  
2     2015   Oct    Monday         Sam Mendes   Male   635.67  False  
3     2012   Jul    Monday  Christopher Nolan   Male   834.94  False  
4     2007   May   Tuesday          Sam Raimi   Male   632.87  False  
...    ...   ...       ...                ...    ...      ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male     0.32  False  
1461  1994   Sep   Tuesday        Kevin Smith   Male     3.12  False  
1462  2009   Aug    Friday           Uwe Boll   Male     0.00  False  
1463  1990   Jul    Friday  Richard Linklater   Male     0.00  False  
1464  1992   Sep    Friday   Robert Rodriguez    NaN     1.82  False  

[1465 rows x 14 columns]

In [ ]:
risky_movies["risky"] == True

0       False
1       False
2       False
3       False
4       False
        ...  
1460    False
1461    False
1462    False
1463    False
1464    False
Name: risky, Length: 1465, dtype: bool

In [ ]:
risky_movies.loc[risky_movies["risky"] == True]

movies_id  budget_mil  popularity  revenue_mil  \
7         43608       200.0         107       586.09   
12        43614       380.0         135      1045.71   
15        43618       200.0          37       310.67   
20        43624       209.0          64       303.03   
24        43630       210.0           3       459.36   
...         ...         ...         ...          ...   
1347      47224         5.0           7         3.26   
1349      47229         5.0           3         4.84   
1351      47233         5.0           6         0.00   
1356      47263        15.0          10         0.00   
1383      47453         3.5           4         0.00   

                                            title  vote_average  vote_count  \
7                               Quantum of Solace           6.1        2965   
12    Pirates of the Caribbean: On Stranger Tides           6.4        4948   
15                                     Robin Hood           6.2        1398   
20                                     Battleship           5.5        2114   
24                          X-Men: The Last Stand           6.3        3525   
...                                           ...           ...         ...   
1347                          The Sweet Hereafter           6.8         103   
1349                         90 Minutes in Heaven           5.4          40   
1351                                Light Sleeper           5.7          15   
1356                           Dying of the Light           4.5         118   
1383                  In the Name of the King III           3.3          19   

      year month        day   director_name gender  profit  risky  
7     2008   Oct   Thursday    Marc Forster   Male  386.09   True  
12    2011   May   Saturday    Rob Marshall   Male  665.71   True  
15    2010   May  Wednesday    Ridley Scott   Male  110.67   True  
20    2012   Apr  Wednesday      Peter Berg   Male   94.03   True  
24    2006   May  Wednesday    Brett Ratner   Male  249.36   True  
...    ...   ...        ...             ...    ...     ...    ...  
1347  1997   May  Wednesday     Atom Egoyan   Male   -1.74   True  
1349  2015   Sep     Friday  Michael Polish   Male   -0.16   True  
1351  1992   Aug     Friday   Paul Schrader    NaN   -5.00   True  
1356  2014   Dec   Thursday   Paul Schrader    NaN  -15.00   True  
1383  2013   Dec     Friday        Uwe Boll   Male   -3.50   True  

[131 rows x 14 columns]

In [ ]:
risky_movies.loc[risky_movies["risky"] == True]["title"].unique()

# These are the all unique risky movies

array(['Quantum of Solace', 'Pirates of the Caribbean: On Stranger Tides',
       'Robin Hood', 'Battleship', 'X-Men: The Last Stand',
       'Terminator Salvation', 'World War Z', 'Hugo', 'Waterworld',
       'Alexander', 'Thor', 'Pan', 'Die Another Day', 'Lethal Weapon 4',
       'The Incredible Hulk', 'Clash of the Titans', 'The 13th Warrior',
       'Batman & Robin', 'Fantastic 4: Rise of the Silver Surfer',
       "Dante's Peak", 'The Aviator', 'Town & Country', 'Little Fockers',
       'End of Days', 'Cloud Atlas', 'The Flowers of War',
       'The Stepford Wives', 'Cutthroat Island', 'Mission to Mars',
       'Mission: Impossible', "Charlotte's Web", 'Flubber',
       'Lost in Space', 'The Brothers Grimm', 'Thirteen Days',
       'Looney Tunes: Back in Action', 'A Sound of Thunder',
       'Exorcist: The Beginning', "Charlie Wilson's War", 'Driven',
       'Die Hard 2', 'Doom', 'The Legend of Hercules', 'The Score',
       'Dreamcatcher', 'Cliffhanger', 'Primary Colors', 'Proof 

In [ ]:
risky_movies.loc[risky_movies["risky"] == True]["title"].nunique() # Total number of unique risky movies

131

In [ ]:
df.shape

(1465, 13)

* Out of 1,465 movies, **131** movies stand out as the risky movies.

## Get the movies released only on Saturdays or Sundays

In [ ]:
df.loc[(df["day"] == "Saturday") | (df["day"] == "Sunday")]

movies_id  budget_mil  popularity  revenue_mil  \
1         43598      300.00         139       961.00   
12        43614      380.00         135      1045.71   
40        43656      200.00          45       769.65   
51        43670      175.00          27       173.00   
69        43709      155.00          39       167.30   
...         ...         ...         ...          ...   
1442      48185        0.86           3         2.45   
1444      48204        0.31           1         1.28   
1447      48231        0.00           3         0.00   
1456      48321        0.01          20         7.00   
1459      48359        0.00           2         0.00   

                                            title  vote_average  vote_count  \
1        Pirates of the Caribbean: At World's End           6.9        4500   
12    Pirates of the Caribbean: On Stranger Tides           6.4        4948   
40                                           2012           5.6        4903   
51                                  Evan Almighty           5.3        1151   
69                                      Alexander           5.6         927   
...                                           ...           ...         ...   
1442                      My Beautiful Laundrette           6.6          51   
1444                                  Tumbleweeds           6.2          11   
1447                         Better Luck Tomorrow           6.5          27   
1456                                   Eraserhead           7.5         485   
1459                            George Washington           6.4          36   

      year month       day       director_name gender  profit  
1     2007   May  Saturday      Gore Verbinski   Male  661.00  
12    2011   May  Saturday        Rob Marshall   Male  665.71  
40    2009   Oct  Saturday     Roland Emmerich   Male  569.65  
51    2007   Jun  Saturday         Tom Shadyac   Male   -2.00  
69    2004   Nov    Sunday        Oliver Stone   Male   12.30  
...    ...   ...       ...                 ...    ...     ...  
1442  1985   Sep  Saturday      Stephen Frears    NaN    1.59  
1444  1999   Nov    Sunday      Gavin O'Connor   Male    0.97  
1447  2002   Jan  Saturday          Justin Lin    NaN    0.00  
1456  1977   Mar  Saturday         David Lynch   Male    6.99  
1459  2000   Oct    Sunday  David Gordon Green   Male    0.00  

[93 rows x 13 columns]

In [ ]:
df.loc[(df["day"] == "Saturday") | (df["day"] == "Sunday")]["title"].unique()

# These all are the unique movies released on Saturdays or Sundays

array(["Pirates of the Caribbean: At World's End",
       'Pirates of the Caribbean: On Stranger Tides', '2012',
       'Evan Almighty', 'Alexander',
       'Harry Potter and the Goblet of Fire', 'Die Another Day',
       "Legend of the Guardians: The Owls of Ga'Hoole",
       'Gangs of New York', 'The Lovely Bones', 'Tropic Thunder',
       'Spy Game', 'The Fast and the Furious: Tokyo Drift', 'Jumper',
       'Con Air', 'The Edge', 'Thirteen Days', 'A Sound of Thunder',
       'Red Dragon', 'War Horse', 'Zathura: A Space Adventure', '300',
       'The Twilight Saga: New Moon', 'Non-Stop', 'Blood Work',
       'Zoolander 2', 'Extremely Loud & Incredibly Close', 'Gattaca',
       'Spy Kids', 'House of Wax', 'A Good Year', 'Snakes on a Plane',
       'Up in the Air', 'Drag Me to Hell', 'Stranger Than Fiction',
       'Octopussy', 'Freaky Friday', 'The Curse of the Jade Scorpion',
       'Rain Man', 'The Face of an Angel', 'The Omen',
       'Finding Neverland', 'The Hoax', 'BloodRayne', 

In [ ]:
df.loc[(df["day"] == "Saturday") | (df["day"] == "Sunday")]["title"].nunique()

93

In [ ]:
df.shape

(1465, 13)

* Out of 1,465 movies, only **93** movies got released on Saturdays or Sundays.

_____ THE END _______